src/main/java/org/prebid/server/bidder/adkerneladn/AdkernelAdnBidder.java

```diff
@@ -203,27 +204,18 @@ private static BidRequest createBidRequest(BidRequest preBidRequest, List<Imp> i
         return bidRequestBuilder.build();
     }
 
-    /**
-     * Builds endpoint url based on adapter-specific pub settings from imp.ext.
-     */
-    private static String buildEndpoint(ExtImpAdkernelAdn impExt, String endpointUrl) {
-        final String updatedEndpointUrl;
+    private String buildEndpoint(ExtImpAdkernelAdn impExt) {
+        final String impHost = impExt.getHost();
+        final String host = StringUtils.isNotBlank(impHost) ? impHost : DEFAULT_DOMAIN;
 
-        if (impExt.getHost() != null) {
-            final URL url;
-            try {
-                url = new URL(endpointUrl);
-            } catch (MalformedURLException e) {
-                throw new PreBidException(
-                        String.format("Error occurred while parsing AdkernelAdn endpoint url: %s", endpointUrl), e);
-            }
-            final String currentHostAndPort = url.getHost() + (url.getPort() == -1 ? "" : ":" + url.getPort());
-            updatedEndpointUrl = endpointUrl.replace(currentHostAndPort, impExt.getHost());
-        } else {
-            updatedEndpointUrl = endpointUrl;
-        }
+        return endpointUrl
+                .replace(URL_HOST_MACRO, host)
+                .replace(URL_PUBLISHER_ID_MACRO, impExt.getPubId().toString());
```

Comment:
May pubId be null?

Reply:
if (adkernelAdnExt.getPubId() == null || adkernelAdnExt.getPubId() < 1) {
        throw new PreBidException(String.format("Invalid pubId value. Ignoring imp id=%s", imp.getId()));
    }
    This is from parse extImp method, so seems it can not be null